pkg/clustermesh/types/addressing.go

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package types
+
+import (
+	"fmt"
+	"net/netip"
+	"strconv"
+	"strings"
+)
+
+//
+// In this file, we define types and utilities for cluster-aware
+// addressing which identifies network endpoints using IP address
+// and an optional ClusterID. With this special addressing scheme,
+// we can distinguish network endpoints (e.g. Pods) that have the
+// same IP address, but belong to the different cluster.
+//
+// A "bare" IP address is still a valid identifier because there
+// are cases that endpoints can be identified without ClusterID
+// (e.g. network endpoint has a unique IP address). We can consider
+// this as a special case that ClusterID "doesn't matter". ClusterID
+// 0 is reserved for indicating that.
+//
+
+// AddrCluster is a type that holds a pair of IP and ClusterID.
+// We should use this type as much as possible when we implement
+// IP + Cluster addressing. We should avoid managing IP and ClusterID
+// separately. Otherwise, it is very hard for code readers to see
+// where we are using cluster-aware addressing.
+type AddrCluster struct {
```

Comment:
Nit: `ClusterAddr` would seem a better name for me...

Reply:
`AddrCluster` or `ClusterAddr` reads as it is an address of a cluster, but not an endpoint in a cluster. However, I cannot think of any better than `IPAddr(With|In)Cluster` which might be a bit too long.

Reply:
It is inspired by the naming convention of the `net/netip` package. They name a pair of IP and a port number as `AddrPort`. So I named a pair of IP and a ClusterID as `AddrCluster`.

Rejected candidates
- ClusterIP => we have ClusterIP as a service type and as a VIP assigned to that service type, so very confusing
- IPCluster => ok, but since we are using `netip.Addr`, I thought it is more straightforward to use `Addr`
- AddrClusterID => ok, but a bit long

Reply:
Thanks for the explanation. Yep, then `AddrCluster` :+1: